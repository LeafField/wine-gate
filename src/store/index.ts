import { create } from "zustand";

type User = {
  username: string;
  id: string;
  email: string | undefined;
};

type Menu = "close" | "member" | "free" | "category";

type Store = {
  user: User | null;
  setUser: (user: User) => void;
  logout: () => void;
  menu: Menu;
  setMenu: (menu: Menu) => void;
};

export const useStore = create<Store>((set) => ({
  user: null,
  setUser: (user) => set({ user }),
  logout: () => set({ user: null }),
  menu: "close",
  setMenu: (menu) => set({ menu }),
}));
