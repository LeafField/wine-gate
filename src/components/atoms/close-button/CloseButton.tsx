"use client";
import React, { FC } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faXmark } from "@fortawesome/free-solid-svg-icons";
import { useStore } from "../../../store";

type Props = {
  rightBorder?: boolean;
};

const CloseButton: FC<Props> = ({ rightBorder = true }) => {
  const { setMenu } = useStore();
  const clickhandler = () => {
    setMenu("close");
  };
  return (
    <button
      type="button"
      className={`flex w-full flex-col items-center justify-center border-white ${rightBorder ? "border-r" : ""}`}
      onClick={clickhandler}
      aria-controls="navigation"
      aria-expanded={true}
    >
      <FontAwesomeIcon
        className="text-menu-icon text-white"
        icon={faXmark}
        aria-hidden={"true"}
      />
      <span className="text-menu-title text-white">閉じる</span>
    </button>
  );
};

export default CloseButton;
