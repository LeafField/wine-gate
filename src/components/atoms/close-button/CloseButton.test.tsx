import React from "react";
import { render, screen, renderHook, act } from "@testing-library/react";
import CloseButton from "./CloseButton";
import { useStore } from "../../../store";

describe("CloseButton", () => {
  test("クラッシュせずにレンダリングされる", () => {
    render(<CloseButton />);
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  test('"閉じる"というテキストを含む', () => {
    render(<CloseButton />);
    expect(screen.getByText("閉じる")).toBeInTheDocument();
  });

  test("rightBorderプロパティがtrueの場合、border-rクラスを適用する", () => {
    render(<CloseButton rightBorder={true} />);
    expect(screen.getByRole("button")).toHaveClass("border-r");
  });

  test("rightBorderプロパティがfalseの場合、border-rクラスを適用しない", () => {
    render(<CloseButton rightBorder={false} />);
    expect(screen.getByRole("button")).not.toHaveClass("border-r");
  });

  test("クリックするとsetMenuが実行され、menuの値がcloseになる", () => {
    const { result } = renderHook(() => useStore());
    act(() => {
      result.current.setMenu("member");
    });
    expect(result.current.menu).toBe("member");
    render(<CloseButton />);
    act(() => {
      screen.getByRole("button").click();
    });
    expect(result.current.menu).toBe("close");
  });

  test("スナップショットテスト", () => {
    const { container } = render(<CloseButton />);
    expect(container).toMatchSnapshot();
  });
});
