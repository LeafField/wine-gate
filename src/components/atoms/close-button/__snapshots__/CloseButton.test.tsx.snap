// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloseButton スナップショットテスト 1`] = `
<div>
  <button
    aria-controls="navigation"
    aria-expanded="true"
    class="flex w-full flex-col items-center justify-center border-white border-r"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-xmark text-menu-icon text-white"
      data-icon="xmark"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 384 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
        fill="currentColor"
      />
    </svg>
    <span
      class="text-menu-title text-white"
    >
      閉じる
    </span>
  </button>
</div>
`;
